crates/sargon/src/system/sargon_os/sargon_os_transaction.rs

```diff
@@ -0,0 +1,169 @@
+        {
+            return Err(
+                CommonError::ReservedInstructionsNotAllowedInManifest {
+                    reserved_instructions: execution_summary
+                        .reserved_instructions
+                        .iter()
+                        .map(|i| i.to_string())
+                        .collect(),
+                },
+            );
+        }
+
+        Ok(TransactionToReview {
+            transaction_manifest,
+            execution_summary,
+        })
+    }
+
+    async fn extract_transaction_signers(
+        &self,
+        manifest_summary: ManifestSummary,
+    ) -> Result<IndexSet<PublicKey>> {
+        let accounts =
+            self.profile_state_holder.accounts_on_current_network()?;
+        let personas =
+            self.profile_state_holder.personas_on_current_network()?;
+
+        let account_entities: IndexSet<_> = manifest_summary
+            .addresses_of_accounts_requiring_auth
+            .iter()
+            .filter_map(|address| {
+                accounts
+                    .iter()
+                    .find(|account| account.address == *address)
+                    .map(AccountOrPersona::AccountEntity)
+            })
+            .collect();
+        let persona_entities: IndexSet<_> = manifest_summary
+            .addresses_of_personas_requiring_auth
+            .iter()
+            .filter_map(|address| {
+                personas
+                    .iter()
+                    .find(|persona| persona.address == *address)
+                    .map(AccountOrPersona::PersonaEntity)
+            })
+            .collect();
+        let mut signer_entities: IndexSet<AccountOrPersona> = IndexSet::new();
+        signer_entities.extend(account_entities);
+        signer_entities.extend(persona_entities);
+        Ok(signer_entities
+            .iter()
+            .flat_map(|entity| {
+                let public_keys: IndexSet<PublicKey> = entity
+                    .virtual_hierarchical_deterministic_factor_instances()
+                    .iter()
+                    .map(|factor_instance| {
+                        factor_instance.public_key.public_key
+                    })
+                    .collect();
+                public_keys
+            })
+            .collect())
+    }
+
+    async fn get_transaction_preview(
+        &self,
+        gateway_client: GatewayClient,
+        manifest: TransactionManifest,
+        network_id: NetworkID,
+        message: Message,
+        nonce: Nonce,
+        notary_public_key: PublicKey,
+    ) -> Result<TransactionPreviewResponse> {
+        // Get all transaction signers
+        let signer_public_keys =
+            self.extract_transaction_signers(manifest.summary()).await?;
+        let epoch = gateway_client.current_epoch().await?;
+        let header = TransactionHeader::new(
+            network_id,
+            epoch,
+            Epoch::from(epoch.0 + 10),
```

Comment:
@GhenadieVP Here on Android we add to the current ledger epoch `1` while on iOS `10`. Not sure which one is right.

Reply:
AFAIK, 10 is the right value.

Reply:
FYI `1` is used on Android only for the transaction preview gateway request, while for the submission `10` is being used same as on iOS.